tough/src/io.rs

```diff
@@ -1,133 +1,154 @@
+            Poll::Ready(None) => {
+                let result = &self.digest.clone().finish();
+                if result.as_ref() != self.hash.as_slice() {
+                    let mismatch_err = error::HashMismatchSnafu {
+                        context: self.url.to_string(),
+                        calculated: hex::encode(result),
+                        expected: hex::encode(&self.hash),
+                    }
+                    .build();
+                    return Poll::Ready(Some(Err(TransportError::new_with_cause(
+                        crate::TransportErrorKind::Other,
+                        self.url.clone(),
+                        mismatch_err,
+                    ))));
                 }
-                .fail()?;
             }
-            Ok(size)
-        } else if let Some(digest) = &mut self.digest {
-            digest.update(&buf[..size]);
-            Ok(size)
-        } else {
-            unreachable!();
-        }
+            Poll::Ready(Some(Err(_))) | Poll::Pending => (),
+        };
+
+        poll
     }
 }
 
-pub(crate) struct MaxSizeAdapter<'a> {
-    reader: Box<dyn Read + Send + 'a>,
-    /// How the `max_size` was specified. For example the max size of `root.json` is specified by
-    /// the `max_root_size` argument in `Settings`. `specifier` is used to construct an error
-    /// message when the `MaxSizeAdapter` detects that too many bytes have been read.
-    specifier: &'static str,
+/// Create a new stream from `stream`. The new stream returns an error for the item that exceeds the
+/// total byte count of `max_size`.
+/// * `stream` - The original stream.
+/// * `max_size` - Size limit in bytes.
+/// * `specifier` - Error message to use.
+pub(crate) fn max_size_adapter(
+    stream: TransportStream,
+    url: Url,
     max_size: u64,
-    counter: u64,
+    specifier: &'static str,
+) -> TransportStream {
+    let mut size: u64 = 0;
+    let stream = stream.map(move |chunk| {
+        if let Ok(bytes) = &chunk {
+            size = size.saturating_add(bytes.len().try_into().unwrap_or(u64::MAX));
+        }
+        if size > max_size {
+            let size_err = error::MaxSizeExceededSnafu {
+                max_size,
+                specifier,
+            }
+            .build();
+            return Err(TransportError::new_with_cause(
+                crate::TransportErrorKind::Other,
+                url.clone(),
+                size_err,
+            ));
+        }
+        chunk
+    });
+
+    stream.boxed()
 }
 
-impl<'a> MaxSizeAdapter<'a> {
-    pub(crate) fn new(
-        reader: Box<dyn Read + Send + 'a>,
-        specifier: &'static str,
-        max_size: u64,
-    ) -> Self {
-        Self {
-            reader,
-            specifier,
-            max_size,
-            counter: 0,
-        }
-    }
+/// Async analogue of `std::path::Path::is_file`
+pub async fn is_file(path: impl AsRef<Path>) -> bool {
+    metadata(path).await.map(|m| m.is_file()).unwrap_or(false)
```

Comment:
nit: slightly clearer that this is a tokio function.

```suggestion
    fs::metadata(path).await.map(|m| m.is_file()).unwrap_or(false)
```